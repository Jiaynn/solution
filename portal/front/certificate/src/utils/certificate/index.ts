/*
 * @file component certificate utils
 * @author Yao Jingtian <yncst233@gmail.com>
 */

import { sortBy, last, without, uniq, values } from 'lodash'

import { OrderType } from '../../apis/ssl'
import {
  sslMap, canBeClosedStatuses, SSLDomainType, CertType,
  OrderStatus, ProductShortName, SslBrand, isDVSSLType
} from '../../constants/ssl'
import { isOEM } from '../../constants/env'
import { isStatusDone, isStatusPending } from './status'

interface ICertBrandInfo {
  brand: SslBrand
  certType: CertType
}

// 根据 productShortName 获取证书品牌、类型
export function shortNameToInfo(productShortName?: ProductShortName): ICertBrandInfo {
  let brand: SslBrand
  let certType: CertType
  Object.keys(sslMap).some((sslBrand: SslBrand) => Object.keys(sslMap[sslBrand]).some((sslType: CertType) => {
    if (sslMap[sslBrand][sslType]!.shortName === productShortName) {
      brand = sslBrand
      certType = sslType
      return true
    }
    if (productShortName && sslMap[sslBrand][sslType]!.domainShortNameMap) {
      const names = values(sslMap[sslBrand][sslType]!.domainShortNameMap)
      if (names.indexOf(productShortName) > -1) {
        brand = sslBrand
        certType = sslType
        return true
      }
    }
    return false
  }))
  return {
    brand: brand!,
    certType: certType!
  }
}

export function getLatestDnsNames(replaceOrders: any[], dnsNames?: string[]): string[] {
  if (!replaceOrders || replaceOrders.length === 0) {
    return dnsNames || []
  }
  const list = sortBy(replaceOrders, 'create_time')
    .filter(order => order.state === 1)
  if (list.length > 0) {
    return last(list).dnsnames.split(',')
  }
  return dnsNames || []
}

export function hasRepeatItem(arr: any[], compareArr?: any[]) {
  if (arr && compareArr) {
    return without(arr.slice(), ...compareArr.slice()).length !== arr.length
  }
  return uniq(arr.slice()).length !== arr.length
}

/**
 * 从后端接口字段 renewable 来判断
 * 历史前端逻辑：已签发/已重颁发 过期前 30 天的非重颁发订单，显示【续费】(若 notAfter <= 0，则根据 providerRenewable 字段)
 */
export function canRenew(renewable: boolean) {
  return isOEM ? false : renewable
}

interface ICanAddDomainParams {
  notAfter: number,
  state: number,
  productType: SSLDomainType,
  orderType: OrderType
}

// 已签发/已重颁发 非重颁发 未过期的多域名、多域名泛域名订单，显示【添加域名】
export function canAddDomain(params: ICanAddDomainParams) {
  const { notAfter, state, productType, orderType } = params
  const now = Date.now()
  if (!state || !isStatusDone(state)) {
    return false
  }
  if ([SSLDomainType.Multiple, SSLDomainType.MultipleWildcard].indexOf(productType) === -1) {
    return false
  }
  if (orderType === OrderType.Replace) {
    return false
  }
  if (!notAfter) {
    return false
  }
  return (
    notAfter <= 0
    || now < notAfter * 1000
  )
}

interface ICanUploadConfirmationParams {
  state: number,
  productShortName: ProductShortName,
  uploadConfirmLetter: boolean
}

// 非 DV 且待确认状态才显示【上传确认函】
export function canUploadConfirmation(params: ICanUploadConfirmationParams) {
  const { state, productShortName, uploadConfirmLetter } = params
  if (!state || !isStatusPending(state)) {
    return false
  }

  const sslInfo = shortNameToInfo(productShortName)

  return !uploadConfirmLetter && !isDVSSLType(sslInfo.certType)
}

export function canClose(state: number): boolean {
  return canBeClosedStatuses.indexOf(state) !== -1
}

// 过期 7 天前的证书都能够部署
export function canCertDeploy(notAfter: number): boolean {
  return !isOEM && ((notAfter - 7 * 24 * 3600) > (Date.now() / 1000))
}

// 判断证书类型是否为私有证书
export function isPrivateCert(productShortName: ProductShortName) {
  return !productShortName
}

// 已经完成的订单，并且满足证书部署条件的订单能够部署
export function canOrderDeploy(notAfter: number, state: OrderStatus): boolean {
  return isStatusDone(state) && canCertDeploy(notAfter)
}

// 2 年的续期证书需要补充文案
export function getCertYearTipsByYear(year: number): string {
  return year === 2 ? '续期证书' : ''
}

// 自动续期的证书、订单需要补充文案
export function getCertYearTipsByAutoRenew(autoRenew: boolean): string {
  return autoRenew ? '续期证书' : ''
}
