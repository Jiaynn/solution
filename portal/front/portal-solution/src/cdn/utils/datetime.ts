import moment from 'moment'

export function isLastDayOfMonth() {
  return moment().endOf('month').format('MM-DD') === moment().format('MM-DD')
}
