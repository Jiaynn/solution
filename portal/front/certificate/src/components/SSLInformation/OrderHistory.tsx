/**
 * @file description OrderHistory for SSLInformation
 * @author [Yao Jingtian] [yncst233@gmail.com]
 */

import React from 'react'
import { observable, action, computed, makeObservable } from 'mobx'
import { observer } from 'mobx-react'
import { without, sortBy, last } from 'lodash'

import { useInjection } from 'qn-fe-core/di'
import Table from 'react-icecream/lib/table'
import Collapse from 'react-icecream/lib/collapse'
import Button from 'react-icecream/lib/button'
import Modal from 'react-icecream/lib/modal'

import { RouterStore } from 'portal-base/common/router'
import { ToasterStore } from 'portal-base/common/toaster'
import { Loadings } from 'portal-base/common/loading'
import { CommonApiException } from 'portal-base/common/apis/common'

import { orderStatusTextMap, orderStatusNameMap, OrderStatus } from '../../constants/ssl'
import SslApis, { IHistoryOrder, IOrderAuth } from '../../apis/ssl'
import { getLatestDnsNames, canClose } from '../../utils/certificate'
import { isStatusPaying } from '../../utils/certificate/status'
import { payForOrder, humanizeTime } from '../../utils/base'
import { basename } from '../../constants/app'
import { AuthMethodType, DnsRecordType, dnsRecordTypeTextMap } from '../../constants/domain'
import { humanizeAuthKey } from '../../transforms/domain'
import AuthValueItem from '../common/AuthValueItem'

const Column = Table.Column
const Panel = Collapse.Panel

export enum OperationType {
  CertDetail = 'certDetail',
  Pay = 'pay',
  Close = 'close',
  Auth = 'auth'
}

interface IOperationInfo {
  operation: OperationType,
  id: string,
  tradeid: string,
  auth?: {
    authMethod: AuthMethodType
    authArray: IOrderAuth[]
    recordType?: DnsRecordType
  }
  isPaid?: boolean
}

export interface IOrderHistoryProps {
  orderid: string,
  originDomains: string[],
  sslBrand: string,
  updateParentDnsNames: (data: string[]) => any
  updateParentCertID: (data: string) => any,
  orders?: any[]
}

type OrderHistoryInnerProps = IOrderHistoryProps & {
  routerStore: RouterStore
  sslApis: SslApis
  toasterStore: ToasterStore
}

@observer
export class OrderHistory extends React.Component<OrderHistoryInnerProps> {
  constructor(props: OrderHistoryInnerProps) {
    super(props)
    makeObservable(this)
    ToasterStore.bindTo(this, props.toasterStore)
    this.fetchOrderHistory = this.fetchOrderHistory.bind(this)
    this.updateReplaceOrders = this.updateReplaceOrders.bind(this)
    this.handleOperation = this.handleOperation.bind(this)
    this.handleCloseOrder = this.handleCloseOrder.bind(this)
    this.closeOrder = this.closeOrder.bind(this)
    this.getAddonDomains = this.getAddonDomains.bind(this)
  }

  loadings = new Loadings('load')
  @computed get isLoading() {
    return this.loadings.isLoading('load')
  }

  @observable orders = this.props.orders || []

  @action updateReplaceOrders(data: IHistoryOrder[]) {
    if (!data) {
      this.orders = []
      return
    }
    this.orders = sortBy(data, 'create_time')
    if (this.orders.length > 0) {
      const latestDnsNames = getLatestDnsNames(this.orders, this.props.originDomains)
      this.props.updateParentDnsNames(latestDnsNames)
      const successOrders = this.orders.filter(order => order.certID && order.certID !== '')
      if (successOrders.length === 0) {
        return
      }
      const lastestCertID = last(successOrders).certID
      this.props.updateParentCertID(lastestCertID)
    }
  }

  @ToasterStore.handle()
  fetchOrderHistory() {
    const result = this.props.sslApis.fetchReplaceOrders(this.props.orderid).then(
      list => this.updateReplaceOrders(list)
    )
    return this.loadings.promise('load', result)
  }

  @ToasterStore.handle('订单关闭成功！')
  closeOrder(orderId: string) {
    return this.props.sslApis.closeSSLOrder(orderId).catch(
      error => {
        if (error instanceof CommonApiException && error.code === 400609) {
          throw error.withMessage('重颁发订单确认中，暂不可关闭')
        }
        return Promise.reject(error)
      }
    )
  }

  handleCloseOrder(orderId: string) {
    this.closeOrder(orderId).then(
      () => this.fetchOrderHistory()
    )
  }

  handleOperation(operationInfo: IOperationInfo) {
    const { routerStore } = this.props
    const operation = operationInfo.operation
    switch (operation) {
      case 'certDetail': {
        routerStore.push(`${basename}/ssl/detail/${operationInfo.id}/cert`)
        break
      }
      case 'pay': {
        payForOrder({ orderId: operationInfo.id, tradeOrderId: operationInfo.tradeid }, routerStore)
        break
      }
      case 'close': {
        const orderid = operationInfo.id
        const tip = (
          operationInfo.isPaid
          ? '申请退款后，该订单将被关闭，退款金额将返还至您的七牛账户，如仍需重颁发，请重新申请。确认要退款吗？'
          : '关闭订单后，如仍需重颁发，请重新申请。确认关闭订单吗？'
        )
        Modal.confirm({
          content: tip,
          onOk: () => this.handleCloseOrder(orderid)
        })
        break
      }
      case 'auth': {
        if (!operationInfo.auth || !operationInfo.auth.authArray) {
          Modal.warning({
            title: '暂无验证信息，请提交工单联系技术支持！'
          })
          return
        }

        const authMethod = operationInfo.auth.authMethod
        const recordType = operationInfo.auth.recordType || DnsRecordType.Txt

        // 这部分 forEach 的过程也可以考虑用 lodash.flatten + map 或者 lodash.flatMap 来做
        const dnsInfoList: React.ReactNode[] = []
        operationInfo.auth.authArray.forEach(({ AuthKey: authKey, AuthValue: authValue }) => dnsInfoList.push(
          <div key={authKey}>{humanizeAuthKey(authKey, authMethod, recordType)}</div>,
          <AuthValueItem key={authValue} authValue={authValue} authMethod={authMethod} recordType={recordType} />
        ))

        Modal.info({
          title: '验证信息',
          width: 740,
          content: (
            <div className="validate-info">
              <div>
                验证类型：{ authMethod === AuthMethodType.Dns ? `DNS(${dnsRecordTypeTextMap[recordType]})` : authMethod }
                {/* 这边使用 <a> 而不是 <Link> 是因为 Modal.info 不能为 content 提供 context */}
                <a href="https://developer.qiniu.com/ssl/manual/3667/ssl-certificate-of-free-dns-validation-guide" target="_blank" rel="noopener" style={{ marginLeft: '10px' }}>(配置指南)</a>
              </div>
              <div className="dns-info">
                { dnsInfoList }
              </div>
            </div>
          )
        })
        break
      }
      default:
    }
  }

  getAddonDomains(dnsnames: string, index: number) {
    const now = dnsnames ? dnsnames.split(',') : []
    if (index === 0) {
      const prevDomains = this.props.originDomains
      return prevDomains.length > 0 ? without(now, ...prevDomains) : now
    }
    const successPrevs = this.orders.slice(0, index).filter(
      order => order.state === 1
    )
    const prev = (
      successPrevs.length > 0
      ? last(successPrevs).dnsnames.split(',')
      : this.props.originDomains
    )
    return prev.length > 0 ? without(now, ...prev) : now
  }

  componentDidMount() {
    if (!this.props.orders) {
      this.fetchOrderHistory()
    }
  }

  render() {
    if (!this.orders || !this.orders.length) {
      return null
    }

    return (
      <Collapse defaultActiveKey={['1']}>
        <Panel header="添加域名记录" key="1">
          <Table key="orderTable"
            dataSource={this.orders.slice()}
            pagination={false}
            rowKey={(record: any) => record.orderid}
            loading={this.isLoading}
          >
            <Column
              title="#"
              dataIndex="orderid"
              key="orders.orderid"
              render={(_text, _record, index) => `${index + 1}`}
            />
            <Column
              title="修改时间"
              dataIndex="create_time"
              key="orders.create_time"
              render={create_time => humanizeTime(create_time)}
            />
            <Column
              title="添加域名"
              dataIndex="dnsnames"
              key="orders.dnsnames"
              render={(text, _record, index) => {
                const addon = this.getAddonDomains(text, index)
                if (addon.length === 0) {
                  return '未修改'
                }
                return (
                  <div>{ addon.map(domain => <div key={domain}>{domain}</div>) }</div>
                )
              }}
            />
            <Column<any>
              title="状态"
              dataIndex="state"
              key="orders.state"
              render={(state: OrderStatus, record) => {
                if (state === OrderStatus.Closed) {
                  return <span className="status-closed">{record.isPaid ? '已退费' : '已关闭'}</span>
                }
                return <span className={`status-${orderStatusNameMap[state]}`}>{orderStatusTextMap[state]}</span>
              }}
            />
            <Column<any>
              title="操作"
              key="orders.operations"
              render={(_text, record, _index) => {
                const operations = []
                // 显示 支付 操作
                if (isStatusPaying(record.state)) {
                  operations.push({ to: OperationType.Pay, text: '支付' })
                }

                // DV 证书、已付款，显示 验证信息
                if (this.props.sslBrand === 'TrustAsia' && !isStatusPaying(record.state)) {
                  operations.push({ to: OperationType.Auth, text: '验证信息' })
                }

                // 关闭订单 / 申请退款
                if (canClose(record.state)) {
                  operations.push({
                    to: OperationType.Close,
                    text: record.isPaid ? '申请退款' : '关闭订单',
                    isPaid: record.isPaid
                  })
                }

                const list = operations.map(operation => {
                  const operationInfo: IOperationInfo = {
                    id: record.orderid,
                    operation: operation.to,
                    tradeid: record.trade_order_id
                  }
                  if (operation.to === OperationType.Auth) {
                    operationInfo.auth = {
                      authMethod: record.auth_method,
                      authArray: record.auth_array,
                      recordType: record.record_type
                    }
                  }
                  return (
                    <a href="javascript:;" key={operation.to}>
                      <Button
                        key={`${record.orderid}.${operation.to}`}
                        type="link"
                        onClick={() => this.handleOperation(operationInfo)}
                      >
                        {operation.text}
                      </Button>
                    </a>
                  )
                })
                return (
                  <span>
                    {list}
                  </span>
                )
              }}
            />
          </Table>
        </Panel>
      </Collapse>
    )
  }
}

export default observer(function _OrderHistory(props: IOrderHistoryProps) {
  const sslApis = useInjection(SslApis)
  const routerStore = useInjection(RouterStore)
  const toasterStore = useInjection(ToasterStore)
  return <OrderHistory {...props} sslApis={sslApis} routerStore={routerStore} toasterStore={toasterStore} />
})
