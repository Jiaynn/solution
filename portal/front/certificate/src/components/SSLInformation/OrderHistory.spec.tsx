/**
 * @file test cases for component OrderHistory in SSLInformation
 * @author [Yao Jingtian] [yncst233@gmail.com]
 */
import React from 'react'

import { RendererUtils as Renderer } from '../../utils/test'

import OrderHistory from './OrderHistory'

const orderId = '123'
const originDomains = [
  'test-1.qiniu.io',
  'test-1-1.qiniu.io',
  'test-1-2.qiniu.io'
]
const orders = [{
  orderid: '599d67e8f9537d17fc000001',
  state: 1,
  reject_reason: '',
  dnsnames: 'test-1-3.qiniu.io,test-1-1.qiniu.io,*.test-1-4.qiniu.io,*.test-1-5.qiniu.io,*.test-1-6.qiniu.io,test-1-2.qiniu.io',
  certID: '599e3d5c0c3ca4d80e000001',
  trade_order_id: 'be341249df108cb23c312ae62b6565cd',
  create_time: 1503487976
}, {
  orderid: '599d3c35f9537d241f000001',
  state: 15,
  reject_reason: '',
  dnsnames: '*.test-1-8.qiniu.io,*.test-1-9.qiniu.io,test-1-3.qiniu.io,test-1-1.qiniu.io,test-1-2.qiniu.io,*.test-1-7.qiniu.io',
  certID: '',
  trade_order_id: '3dcaf04c357c577a857f3ffadc555f9b',
  create_time: 1503476789
}, {
  orderid: '599c0675f9537d3f4c000004',
  state: 2,
  reject_reason: '',
  dnsnames: '*.test-1-5.qiniu.io,*.test-1-6.qiniu.io,test-1-1.qiniu.io,test-1-2.qiniu.io,test-1-3.qiniu.io,*.test-1-4.qiniu.io',
  certID: '',
  trade_order_id: '10b4945abe2e627db646b3c5226a4e50',
  create_time: 1503397493
}, {
  orderid: '599bd168f9537d2217000001',
  state: 1,
  reject_reason: '',
  dnsnames: 'test-1-1.qiniu.io,test-1-2.qiniu.io,test-1-3.qiniu.io',
  certID: '599bf377f9537d2217000002',
  trade_order_id: '',
  create_time: 1503383912
}
]

it('renders replace order info correctly', () => {
  const renderer = new Renderer()
  const updateDns = (data: string[]) => console.log(data)
  const updateCertID = (data: string) => console.log(data)
  const tree = renderer.create(
    <OrderHistory
      orderid={orderId}
      sslBrand="DigiCert"
      originDomains={originDomains}
      orders={orders}
      updateParentDnsNames={updateDns}
      updateParentCertID={updateCertID} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders only reissue order correctly', () => {
  const renderer = new Renderer()
  const updateDns = (data: string[]) => console.log(data)
  const updateCertID = (data: string) => console.log(data)

  const dvReplaceOrders = [{
    orderid: '599bd168f9537d2217000001',
    state: 1,
    reject_reason: '',
    dnsnames: '',
    certID: '599bf377f9537d2217000002',
    trade_order_id: '',
    create_time: 1503383912,
    auth_method: 'FILE',
    auth_array: [{
      AuthKey: 'test-dv.qbox.net',
      AuthValue: 'abcdedcba'
    }]
  }]
  const tree = renderer.create(
    <OrderHistory
      orderid={orderId}
      sslBrand="TrustAsia"
      originDomains={['test-dv.qbox.net']}
      orders={dvReplaceOrders}
      updateParentDnsNames={updateDns}
      updateParentCertID={updateCertID} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
