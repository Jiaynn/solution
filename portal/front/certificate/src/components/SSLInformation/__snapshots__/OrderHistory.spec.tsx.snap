// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders only reissue order correctly 1`] = `
<div
  className="ant-collapse ant-collapse-icon-position-left"
  role={null}
>
  <div
    className="ant-collapse-item ant-collapse-item-active"
  >
    <div
      aria-expanded="true"
      className="ant-collapse-header"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <i
        aria-label="icon: right"
        className="anticon anticon-right ant-collapse-arrow"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          style={
            Object {
              "msTransform": "rotate(90deg)",
              "transform": "rotate(90deg)",
            }
          }
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
          />
        </svg>
      </i>
      添加域名记录
    </div>
    <div
      className="ant-collapse-content ant-collapse-content-active"
      role={null}
    >
      <div
        className="ant-collapse-content-box"
      >
        <div
          className="ant-table-wrapper"
        >
          <div
            className="ant-spin-nested-loading"
          >
            <div
              className="ant-spin-container"
            >
              <div
                className="ant-table ant-table-default ant-table-scroll-position-left"
                style={Object {}}
              >
                <div
                  className="ant-table-content"
                >
                  <div
                    className="ant-table-body"
                    onScroll={[Function]}
                    onWheel={[Function]}
                    style={Object {}}
                  >
                    <table
                      className=""
                      style={Object {}}
                    >
                      <colgroup>
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                      </colgroup>
                      <thead
                        className="ant-table-thead"
                      >
                        <tr
                          style={
                            Object {
                              "height": null,
                            }
                          }
                        >
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  #
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  修改时间
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  添加域名
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  状态
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  操作
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        className="ant-table-tbody"
                      >
                        <tr
                          className="ant-table-row ant-table-row-level-0"
                          data-row-key="599bd168f9537d2217000001"
                          onClick={[Function]}
                          onContextMenu={[Function]}
                          onDoubleClick={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "height": null,
                            }
                          }
                        >
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="ant-table-row-indent indent-level-0"
                              style={
                                Object {
                                  "paddingLeft": "0px",
                                }
                              }
                            />
                            1
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            2017-08-22 14:38:32
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            未修改
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="status-done"
                            >
                              已签发
                            </span>
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span>
                              <a
                                href="javascript:;"
                              >
                                <button
                                  className="ant-btn ant-btn-link"
                                  onClick={[Function]}
                                  type="button"
                                >
                                  <span>
                                    验证信息
                                  </span>
                                </button>
                              </a>
                            </span>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders replace order info correctly 1`] = `
<div
  className="ant-collapse ant-collapse-icon-position-left"
  role={null}
>
  <div
    className="ant-collapse-item ant-collapse-item-active"
  >
    <div
      aria-expanded="true"
      className="ant-collapse-header"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <i
        aria-label="icon: right"
        className="anticon anticon-right ant-collapse-arrow"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          style={
            Object {
              "msTransform": "rotate(90deg)",
              "transform": "rotate(90deg)",
            }
          }
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
          />
        </svg>
      </i>
      添加域名记录
    </div>
    <div
      className="ant-collapse-content ant-collapse-content-active"
      role={null}
    >
      <div
        className="ant-collapse-content-box"
      >
        <div
          className="ant-table-wrapper"
        >
          <div
            className="ant-spin-nested-loading"
          >
            <div
              className="ant-spin-container"
            >
              <div
                className="ant-table ant-table-default ant-table-scroll-position-left"
                style={Object {}}
              >
                <div
                  className="ant-table-content"
                >
                  <div
                    className="ant-table-body"
                    onScroll={[Function]}
                    onWheel={[Function]}
                    style={Object {}}
                  >
                    <table
                      className=""
                      style={Object {}}
                    >
                      <colgroup>
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                        <col
                          style={
                            Object {
                              "minWidth": undefined,
                              "width": undefined,
                            }
                          }
                        />
                      </colgroup>
                      <thead
                        className="ant-table-thead"
                      >
                        <tr
                          style={
                            Object {
                              "height": null,
                            }
                          }
                        >
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  #
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  修改时间
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  添加域名
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  状态
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                          <th
                            className=""
                          >
                            <span
                              className="ant-table-header-column"
                            >
                              <div>
                                <span
                                  className="ant-table-column-title"
                                >
                                  操作
                                </span>
                                <span
                                  className="ant-table-column-sorter"
                                />
                              </div>
                            </span>
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        className="ant-table-tbody"
                      >
                        <tr
                          className="ant-table-row ant-table-row-level-0"
                          data-row-key="599d67e8f9537d17fc000001"
                          onClick={[Function]}
                          onContextMenu={[Function]}
                          onDoubleClick={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "height": null,
                            }
                          }
                        >
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="ant-table-row-indent indent-level-0"
                              style={
                                Object {
                                  "paddingLeft": "0px",
                                }
                              }
                            />
                            1
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            2017-08-23 19:32:56
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <div>
                              <div>
                                test-1-3.qiniu.io
                              </div>
                              <div>
                                *.test-1-4.qiniu.io
                              </div>
                              <div>
                                *.test-1-5.qiniu.io
                              </div>
                              <div>
                                *.test-1-6.qiniu.io
                              </div>
                            </div>
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="status-done"
                            >
                              已签发
                            </span>
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span />
                          </td>
                        </tr>
                        <tr
                          className="ant-table-row ant-table-row-level-0"
                          data-row-key="599d3c35f9537d241f000001"
                          onClick={[Function]}
                          onContextMenu={[Function]}
                          onDoubleClick={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "height": null,
                            }
                          }
                        >
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="ant-table-row-indent indent-level-0"
                              style={
                                Object {
                                  "paddingLeft": "0px",
                                }
                              }
                            />
                            2
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            2017-08-23 16:26:29
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <div>
                              <div>
                                *.test-1-8.qiniu.io
                              </div>
                              <div>
                                *.test-1-9.qiniu.io
                              </div>
                              <div>
                                *.test-1-7.qiniu.io
                              </div>
                            </div>
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="status-closed"
                            >
                              已关闭
                            </span>
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span />
                          </td>
                        </tr>
                        <tr
                          className="ant-table-row ant-table-row-level-0"
                          data-row-key="599c0675f9537d3f4c000004"
                          onClick={[Function]}
                          onContextMenu={[Function]}
                          onDoubleClick={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "height": null,
                            }
                          }
                        >
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="ant-table-row-indent indent-level-0"
                              style={
                                Object {
                                  "paddingLeft": "0px",
                                }
                              }
                            />
                            3
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            2017-08-22 18:24:53
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            未修改
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="status-pending"
                            >
                              待确认
                            </span>
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span>
                              <a
                                href="javascript:;"
                              >
                                <button
                                  className="ant-btn ant-btn-link"
                                  onClick={[Function]}
                                  type="button"
                                >
                                  <span>
                                    关闭订单
                                  </span>
                                </button>
                              </a>
                            </span>
                          </td>
                        </tr>
                        <tr
                          className="ant-table-row ant-table-row-level-0"
                          data-row-key="599bd168f9537d2217000001"
                          onClick={[Function]}
                          onContextMenu={[Function]}
                          onDoubleClick={[Function]}
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "height": null,
                            }
                          }
                        >
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="ant-table-row-indent indent-level-0"
                              style={
                                Object {
                                  "paddingLeft": "0px",
                                }
                              }
                            />
                            4
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            2017-08-22 14:38:32
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            未修改
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span
                              className="status-done"
                            >
                              已签发
                            </span>
                          </td>
                          <td
                            className=""
                            onClick={[Function]}
                          >
                            <span />
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
